library/core/src/iter/adapters/step_by.rs

```diff
@@ -233,3 +225,96 @@ where
 // StepBy can only make the iterator shorter, so the len will still fit.
 #[stable(feature = "iterator_step_by", since = "1.28.0")]
 impl<I> ExactSizeIterator for StepBy<I> where I: ExactSizeIterator {}
+
+trait SpecRange: Iterator {
+    fn spec_fold<Acc, F>(self, acc: Acc, f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc;
+}
+
+impl<I> SpecRange for StepBy<I>
+where
+    I: Iterator,
+{
+    default fn spec_fold<Acc, F>(mut self, mut acc: Acc, mut f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc,
+    {
+        #[inline]
+        fn nth<I: Iterator>(iter: &mut I, step: usize) -> impl FnMut() -> Option<I::Item> + '_ {
+            move || iter.nth(step)
+        }
+
+        if self.first_take {
+            self.first_take = false;
+            match self.iter.next() {
+                None => return acc,
+                Some(x) => acc = f(acc, x),
+            }
+        }
+        from_fn(nth(&mut self.iter, self.step)).fold(acc, f)
+    }
+}
+
+macro_rules! spec_int_ranges {
+    ($($t:ty)*) => ($(
+        impl SpecRange for StepBy<Range<$t>> {
+            #[inline]
+            fn spec_fold<Acc, F>(self, mut acc: Acc, mut f: F) -> Acc
+            where
+                F: FnMut(Acc, Self::Item) -> Acc
+            {
+                use crate::convert::TryFrom;
+
+                let mut range = self.iter;
+
+                if !self.first_take {
+                    // `next()` implements `increment-by-step, get-and-increment-by-one`, leaving
+                    // range.start set to one element after the previously returned one.
+                    // This fold uses `get, increment-by-step+1` which leaves the value
+                    // set to one for the next iteration.
+                    //
+                    // Here we get back into the desired state if the iterator
+                    // has been partially advanced
+
+                    // Don't care about overshooting the range end since it'll be dropped at the end
+                    // of the function
+                    range.start = range.start.saturating_add(<$t>::try_from(self.step).unwrap_or(<$t>::MAX));
+                }
+
+                // if a step size larger than the type has been specified fall back to
+                // t::MAX which will never yield more than 1 item.
+                let step = <$t>::try_from(self.step + 1).unwrap_or(<$t>::MAX);
```

Comment:
suggestion: comment that `step + 1` cannot overflow since we subtracted it when creating the iterator.